xtractmime/_utils.py

```diff
@@ -0,0 +1,252 @@
+from struct import unpack
+from typing import Tuple, Union
+
+from xtractmime import _is_match_mime_pattern
+from xtractmime._patterns import (
+    IMAGE_PATTERNS,
+    AUDIO_VIDEO_PATTERNS,
+    FONT_PATTERNS,
+    ARCHIVE_PATTERNS,
+)
+
+sample_rates = [44100, 48000, 32000]
+mp25_rates = [
+    0,
+    8000,
+    16000,
+    24000,
+    32000,
+    40000,
+    48000,
+    56000,
+    64000,
+    80000,
+    96000,
+    112000,
+    128000,
+    144000,
+    160000,
+]
+mp3_rates = [
+    0,
+    32000,
+    40000,
+    48000,
+    56000,
+    64000,
+    80000,
+    96000,
+    112000,
+    128000,
+    160000,
+    192000,
+    224000,
+    256000,
+    320000,
+]
+
+
+def _is_mp4_signature(input_bytes: bytes) -> bool:
+    input_size = len(input_bytes)
+    if input_size < 12:
+        return False
+
+    box_size = unpack(">I", input_bytes[0:4])[0]
+
+    if (input_size < box_size) or (box_size % 4 != 0):
+        return False
+
+    if input_bytes[4:8] != b"ftyp":
+        return False
+
+    if input_bytes[8:11] == b"mp4":
+        return True
+
+    bytes_read = 16
+    while bytes_read < box_size:
+        if input_bytes[bytes_read : bytes_read + 3] == b"mp4":
+            return True
+        bytes_read += 4
+
+    return False
+
+
+def _parse_vint(input_bytes: bytes, input_size: int, index: int) -> int:
```

Comment:
Given the implementation change, this may be a more appropriate name:

```suggestion
def parse_vint_number_size(input_bytes: bytes, input_size: int, index: int) -> int:
```

I’m also thinking that a single `input_bytes` parameter would make more sense, provided that call `_parse_vint(input_bytes, input_size, index)` gets replaced by something like `_parse_vint(memoryview(input_bytes)[index:])`. In the implementation, `index` will always be 0, 
and `len(input_bytes)` should be a valid replacement for `input_size`, and I think it’s actually fixing the algorithm. This also applies to similar functions below; if the algorithm can be implemented with just a memoryview as input, it’s probably better to do that.

Finally, in these functions I think it would also make sense to precede them with a code comment with a link to the snapshot of the standard they are modeled from, for future reference. And, in this specific case, you could explain why the difference with the standard.

Reply:
I think you are talking about the `_match_mp3_header` function under `_is_mp3_non_ID3_signature` where index is initially 0 but it changes later for the second call of the function. 

In case of `_parse_vint`, there too  index keeps on changing. Also, I think it would not be good to use `memoryview` for the _parse_vint function because of the bitwise AND operation associated with it in `_parse_vint` as below.
```
def parse_vint_number_size(input_bytes: bytes, index: int) -> int:
    input_size = len(input_bytes)
    mask = 128
    max_vint_size = 8
    number_size = 1

    while number_size < max_vint_size and number_size < input_size:
        if input_bytes[index] & mask != 0:
```

Reply:
> think it would not be good to use `memoryview` for the _parse_vint function because of the bitwise AND operation associated with it in `_parse_vint` as below.

Could you elaborate why? `bytes` and `memoryview` seem to work the same as far as bitwise operators is concerned:

```python
>>> value = b'\x79\x80'
>>> value_reference = memoryview(value)
>>> value[0] & 128
0
>>> value_reference[0] & 128
0
>>> value[1] & 128
128
>>> value_reference[1] & 128
128
```

Reply:
Sorry, the memoryview is working. I was testing it wrong earlier 😅

Reply:
So, the function is definition looks like `def parse_vint_number_size(input_bytes: memoryview, index: int) -> int:` now and the function call will be `parse_vint_number_size(memoryview(input_bytes), index)`

Reply:
My suggestion is to remove the `index` parameter as well.

If you build the memory view as `memoryview(input_bytes)[index:]`, then you can pass the memory view _only_ as a parameter, and assume the `index` within the function to be `0`.

Reply:
> If you build the memory view as memoryview(input_bytes)[index:], then you can pass the memory view only as a parameter, and assume the index within the function to be 0.

I believe that will change the `input_size`. Will that be a problem??

Reply:
I might be wrong, but given `max_vint_size` is `8`, and given input_size is only used in `while number_size < max_vint_size and number_size < input_size:`, I think the goal of `input_size` is preventing going beyond the length of `input_bytes` (i.e. do not let `input_bytes[index]` raise an `IndexError`).

So I think setting `input_size` to `len(input_bytes)` without the function, if `input_bytes` in the function comes from `memoryview(input_bytes)[index:]` outside the function, would still do the job.

If I understand the algorithm correctly, this is what it does (in a more Pythonic way):

```python
    mask = 128
    max_vint_size = 8
    limit = min(max_vint_size, input_size)
    first_byte = input_bytes[0]
    for number_size in range(1, limit):
        if first_byte & mask:
            return number_size
        mask >>= 1
    return limit
```

I’m not 100% sure, so I suggest you try to think of a scenario where the results would be different before and after this change, and we can discuss whether that scenario proves that the suggested change breaks thing, or that there was a minor bug in the initial implementation.

Reply:
you are right, the above changes are working, I tried all possible scenarios 👍🏼